<div dir=rtl align=center>

<img src='https://upload.wikimedia.org/wikipedia/fa/thumb/a/a9/Sharif_logo.svg/626px-Sharif_logo.svg.png?20110526112825' alt="SUT logo" width=200 height=200 align=center  >
<br>
<font face="B Yekan">
<font color=0F5298 size=7>
یادگیری ماشین<br>
<font color=2565AE size=5>
دانشکده مهندسی صنایع<br>
<font color=2565AE size=4>
دکتر مهدی شریف زاده <br>
<font color=2565AE size=4>
محمدتقی دهقان نژاد <br>
بهار 1402<br>

<font color=3C99D size=5>
تمرین عملی 2
<br>
SVM 
<br>
100 نمره
    
    
____


### Full Name : Parsa Bakhtiari

### Student Number : 97104118



In [8]:
from sklearn.svm import SVC
from sklearn.svm import LinearSVC
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score
from sklearn.model_selection import GridSearchCV

# Load data

Load the breast cancer dataset from scikit-learn using the following code:


In [2]:
from sklearn.datasets import load_breast_cancer
data = load_breast_cancer()
df_base = pd.DataFrame(data.data, columns=data.feature_names)
df_base['target'] = pd.Series(data.target)

# A. Data Splitting (5 points)

Split the dataset into training and testing sets using a 70/30 split.


In [13]:
train_set, test_set = train_test_split(df_base, test_size=0.3, random_state=42)
train_y = train_set['target']
train_x = train_set.drop(['target'], axis=1)
test_y = test_set['target']
test_x = test_set.drop(['target'], axis=1)
scaler = StandardScaler()
train_x_scaled = scaler.fit_transform(train_x)
test_x_scaled = scaler.fit_transform(test_x)

# B. Linear SVC Model (15 points)

Implement a linear SVM classifier using scikit-learn's LinearSVC. Train the classifier on the training set and evaluate its accuracy on the testing set. {15 points}


In [18]:

linear_svc = Pipeline([
    ('Scaler', StandardScaler()),
    ('LinearSVC', LinearSVC(C=0.01, loss='hinge'))
])
linear_svc.fit(train_x, train_y)   
accuracy = accuracy_score(test_y, linear_svc.predict(test_x))
print("train: ",accuracy_score(train_y, linear_svc.predict(train_x)))
print("test: ",accuracy)


train:  0.9773869346733668
test:  0.9883040935672515


# C. RBF Kernel SVM (15 points)

Implement a kernel SVM classifier using scikit-learn's SVC with the RBF kernel. Train the classifier on the training set and evaluate its accuracy on the testing set. Experiment with different values of the gamma parameter to see how it affects the performance of the classifier. {15 points}

In [20]:
gamma_range = np.logspace(-3, 3, num=7)
param_grid = {'gamma': gamma_range}
# rbf_svc = Pipeline([
#     ('Scaler', StandardScaler()),
#     ('svm_rbf', SVC(kernel='rbf'))
# ])
# rbf_svc.fit(train_x, train_y)

rbf_svc = SVC(kernel='rbf')
grid_search = GridSearchCV(rbf_svc, param_grid, cv=5)
grid_search.fit(train_x_scaled, train_y)
accuracy = accuracy_score(test_y, grid_search.predict(test_x_scaled))
print(f"Best gamma = {grid_search.best_estimator_}")
print("Train accuracy: ",accuracy_score(train_y, grid_search.predict(train_x_scaled)))
print(f"Accuracy = {accuracy}")

Best gamma = SVC(gamma=0.01)
Train accuracy:  0.9698492462311558
Accuracy = 0.9766081871345029


# D. Polynomial Kernel SVM (15 points)

Implement a kernel SVM classifier using scikit-learn's SVC with the polynomial kernel. Train the classifier on the training set and evaluate its accuracy on the testing set. Experiment with different values of the degree and coef0 parameters to see how they affect the performance of the classifier. {15 points}

In [21]:
# poly_svc = Pipeline([
#     ('Scaler', StandardScaler()),
#     ('svm_poly', SVC(kernel='poly', C=5, degree=2, coef0=1))
# ])
# poly_svc.fit(train_x, train_y)
# accuracy = accuracy_score(test_y, poly_svc.predict(test_x))
param_grid = {
    'degree': [2, 3, 4, 5, 6, 7],
    'coef0': [0.0, 0.25, 0.5, 0.75, 1.0]
}
poly_svc = SVC(kernel='poly')
grid_search = GridSearchCV(poly_svc, param_grid, cv=5)
grid_search.fit(train_x_scaled, train_y)
accuracy = accuracy_score(test_y, grid_search.predict(test_x_scaled))
print(f"Best params = {grid_search.best_estimator_}")
print("Train accuracy: ",accuracy_score(train_y, grid_search.predict(train_x_scaled)))
print(f"Accuracy = {accuracy}")


Best params = SVC(coef0=1.0, degree=4, kernel='poly')
Train accuracy:  0.992462311557789
Accuracy = 0.9824561403508771


# E. Comparison (25 points)

Compare the accuracy of the three classifiers implemented in questions 2, 3, and 4. Which classifier performed better and why?(There is no need to code, explain in text **completely**) {25 points}


accuracy of 2: 0.9883040935672515

accuracy of 3: 0.9766081871345029

accuracy of 4: 0.9824561403508771

comparing: it seems Linear SVC predicts test sets more accurately than SVM with poly and RBF kernels. This could have multiple reasons. However, in some cases, a linear boundary may be sufficient to separate the data points. Generally, this occurs when a straight line or a plane can easily separate the data. As a result, a linear classifier like LinearSVC can achieve high accuracy. In such cases, polynominal or rbf kernels can overfit the model and linear models can take over the performance of the model.

# F. Decision Boundary Visualization (25 points)

Visualize the decision boundaries for the classifiers implemented in questions 2, 3, and 4 using matplotlib. Plot the training and testing data as well. {25 points}


In [7]:


# Plot decision boundaries for linear SVM classifier (question 2)
linear_svc.fit(X_train[:, :2], y_train)  # only using first two features for visualization
#TODO
plt.xlabel('Feature 1')
plt.ylabel('Feature 2')
plt.title('Linear SVM Decision Boundary')
plt.show()

# Plot decision boundaries for RBF kernel SVM classifier (question 3)
rbf_svc.fit(X_train[:, :2], y_train)  # only using first two features for visualization
#TODO
plt.xlabel('Feature 1')
plt.ylabel('Feature 2')
plt.title('RBF SVM Decision Boundary')
plt.show()

# Plot decision boundaries for polynomial kernel SVM classifier (question 4)
poly_svc.fit(X_train[:, :2], y_train)  # only using first two features for visualization
#TODO
plt.xlabel('Feature 1')
plt.ylabel('Feature 2')
plt.title('Polynomial SVM Decision Boundary')
plt.show()


NameError: name 'X_train' is not defined